File: SpaceGame/Engine/Library/Quadtree.vb
'http://alopex.li/wiki/QuadTree

#Region "Imports"

'System Imports
Imports System.Collections.Generic

'OpenTK Imports
Imports OpenTK

#End Region

Namespace Isotope.Library
    Public Class QuadTree(Of T)
#Region "Variables & Properties"

        Public Loc As Vector2
        Public Item As T
        Public UpperLeft As QuadTree(Of T)
        Public UpperRight As QuadTree(Of T)
        Public LowerLeft As QuadTree(Of T)
        Public LowerRight As QuadTree(Of T)

#End Region
#Region "Initializers"


        Public Sub New(ByVal loc__1 As Vector2, ByVal i As T)
            Item = i
            Loc = loc__1
            UpperLeft = Nothing
            UpperRight = Nothing
            LowerLeft = Nothing
            LowerRight = Nothing
        End Sub

#End Region
#Region "Main Functions"

        Public Sub Insert(ByVal loc__1 As Vector2, ByVal item As T)
            If loc__1.X > Loc.X Then
                If loc__1.Y > Loc.Y Then
                    If UpperRight Is Nothing Then
                        UpperRight = New QuadTree(Of T)(loc__1, item)
                    Else
                        UpperRight.Insert(loc__1, item)
                    End If
                Else
                    ' loc.Y <= Loc.Y
                    If LowerRight Is Nothing Then
                        LowerRight = New QuadTree(Of T)(loc__1, item)
                    Else
                        LowerRight.Insert(loc__1, item)
                    End If
                End If
            Else
                ' loc.X <= Loc.X
                If loc__1.Y > Loc.Y Then
                    If UpperLeft Is Nothing Then
                        UpperLeft = New QuadTree(Of T)(loc__1, item)
                    Else
                        UpperLeft.Insert(loc__1, item)
                    End If
                Else
                    ' loc.Y <= Loc.Y
                    If LowerLeft Is Nothing Then
                        LowerLeft = New QuadTree(Of T)(loc__1, item)
                    Else
                        LowerLeft.Insert(loc__1, item)
                    End If
                End If
            End If
        End Sub

        Public Function GetWithin(ByVal target As Vector2, ByVal d As Double) As List(Of T)
            Return GetWithin(target, d, New List(Of T)())
        End Function

        ' Returns a list of items within d of the target
        ' Upon consideration, it feels weird not making this tail-recursive.
        Public Function GetWithin(ByVal target As Vector2, ByVal d As Double, ByVal ret As List(Of T)) As List(Of T)
            Dim dsquared As Double = d * d
            ' First, we check and see if the current item is in range
            If GameMath.Vector2DistanceSquared(Loc, target) < dsquared Then
                ret.Add(Item)
            End If

            If target.X + d > Loc.X OrElse target.X > Loc.X Then
                If target.Y + d > Loc.Y OrElse target.Y > Loc.Y Then
                    If UpperRight IsNot Nothing Then
                        UpperRight.GetWithin(target, d, ret)
                    End If
                End If
                If target.Y - d <= Loc.Y OrElse target.Y <= Loc.Y Then
                    If LowerRight IsNot Nothing Then
                        LowerRight.GetWithin(target, d, ret)
                    End If
                End If
            End If
            If target.X - d <= Loc.X OrElse target.X <= Loc.X Then
                If target.Y + d > Loc.Y OrElse target.Y > Loc.Y Then
                    If UpperLeft IsNot Nothing Then
                        UpperLeft.GetWithin(target, d, ret)
                    End If
                End If
                If target.Y - d <= Loc.Y OrElse target.Y <= Loc.Y Then
                    If LowerLeft IsNot Nothing Then
                        LowerLeft.GetWithin(target, d, ret)
                    End If
                End If
            End If

            Return ret
        End Function

#End Region
    End Class
End Namespace

File: SpaceGame/Engine/Library/Quadtree.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
//http://alopex.li/wiki/QuadTree

#region "Imports"

//System Imports

//OpenTK Imports
using OpenTK;

#endregion

namespace Isotope.Library
{
    public class QuadTree<T>
    {
        #region "Variables & Properties"

        public Vector2 Loc;
        public T Item;
        public QuadTree<T> UpperLeft;
        public QuadTree<T> UpperRight;
        public QuadTree<T> LowerLeft;

        public QuadTree<T> LowerRight;
        #endregion
        #region "Initializers"


        public QuadTree(Vector2 loc__1, T i)
        {
            Item = i;
            Loc = loc__1;
            UpperLeft = null;
            UpperRight = null;
            LowerLeft = null;
            LowerRight = null;
        }

        #endregion
        #region "Main Functions"

        public void Insert(Vector2 loc__1, T item)
        {
            if (loc__1.X > Loc.X)
            {
                if (loc__1.Y > Loc.Y)
                {
                    if (UpperRight == null)
                    {
                        UpperRight = new QuadTree<T>(loc__1, item);
                    }
                    else
                    {
                        UpperRight.Insert(loc__1, item);
                    }
                }
                else
                {
                    // loc.Y <= Loc.Y
                    if (LowerRight == null)
                    {
                        LowerRight = new QuadTree<T>(loc__1, item);
                    }
                    else
                    {
                        LowerRight.Insert(loc__1, item);
                    }
                }
            }
            else
            {
                // loc.X <= Loc.X
                if (loc__1.Y > Loc.Y)
                {
                    if (UpperLeft == null)
                    {
                        UpperLeft = new QuadTree<T>(loc__1, item);
                    }
                    else
                    {
                        UpperLeft.Insert(loc__1, item);
                    }
                }
                else
                {
                    // loc.Y <= Loc.Y
                    if (LowerLeft == null)
                    {
                        LowerLeft = new QuadTree<T>(loc__1, item);
                    }
                    else
                    {
                        LowerLeft.Insert(loc__1, item);
                    }
                }
            }
        }

        public List<T> GetWithin(Vector2 target, double d)
        {
            return GetWithin(target, d, new List<T>());
        }

        // Returns a list of items within d of the target
        // Upon consideration, it feels weird not making this tail-recursive.
        public List<T> GetWithin(Vector2 target, double d, List<T> ret)
        {
            double dsquared = d * d;
            // First, we check and see if the current item is in range
            if (GameMath.Vector2DistanceSquared(Loc, target) < dsquared)
            {
                ret.Add(Item);
            }

            if (target.X + d > Loc.X || target.X > Loc.X)
            {
                if (target.Y + d > Loc.Y || target.Y > Loc.Y)
                {
                    if (UpperRight != null)
                    {
                        UpperRight.GetWithin(target, d, ret);
                    }
                }
                if (target.Y - d <= Loc.Y || target.Y <= Loc.Y)
                {
                    if (LowerRight != null)
                    {
                        LowerRight.GetWithin(target, d, ret);
                    }
                }
            }
            if (target.X - d <= Loc.X || target.X <= Loc.X)
            {
                if (target.Y + d > Loc.Y || target.Y > Loc.Y)
                {
                    if (UpperLeft != null)
                    {
                        UpperLeft.GetWithin(target, d, ret);
                    }
                }
                if (target.Y - d <= Loc.Y || target.Y <= Loc.Y)
                {
                    if (LowerLeft != null)
                    {
                        LowerLeft.GetWithin(target, d, ret);
                    }
                }
            }

            return ret;
        }

        #endregion
    }
}
